### PR TITLE
Intermediate CTC + Stochastic depth

### PR DESCRIPTION
**Pull request by jaesong**
This implements [intermediate CTC](https://arxiv.org/abs/2102.03216) with [stochastic depth](https://arxiv.org/abs/1603.09382) for Transformer and Conformer encoders.

`stochastic_depth_rate` indicates the drop probability, similar to the dropout parameter. This might be confusing as the many papers use the notation of keep probability (`= 1 - drop_rate`). But I think it would be better to follow dropout notation.

The return type of Encoder needs to be changed due to intermediate CTC. To minimize the issue, the return type is changed only if `Encoder(intermediate_layers=...)` is given.

I would add some recipes soon.

### CODE REVIEW

**Comment by sw005320** [comment]
Thanks, @jaesong!
This is really cool!

**Comment by sw005320** [comment]
Can you fix some CI errors? (https://travis-ci.org/github/espnet/espnet/builds/773609667#L2756)

**Review by sw005320 — COMMENTED** [review]
The code looks very good to me
- Can you add an integration test?
- Can you prepare a config file in some recipes? If it observes the improvement in that recipe, we can use it as a default config (by changing the symlink).

File: espnet/nets/pytorch_backend/transformer/encoder.py [hunk 1]
```diff
@@ -298,9 +307,23 @@ def forward(self, xs, masks):
...
+        intermediate_outputs = []
+        for layer_idx, encoder_layer in enumerate(self.encoders):
+            xs, masks = encoder_layer(xs, masks)
+
+            if self.intermediate_layers is not None and layer_idx + 1 in self.intermediate_layers:
+                encoder_output = xs
+                # intermediate branches also require normalization.
+                if self.normalize_before:
+                    encoder_output = self.after_norm(encoder_output)
+                intermediate_outputs.append(encoder_output)
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/transformer/encoder.py:320]:
```suggestion
        if self.intermediate_layers is None:
            xs, masks = self.encoders(xs, masks)
        else:
            intermediate_outputs = []
            for layer_idx, encoder_layer in enumerate(self.encoders):
                xs, masks = encoder_layer(xs, masks)
    
                if layer_idx + 1 in self.intermediate_layers:
                    encoder_output = xs
                    # intermediate branches also require normalization.
                    if self.normalize_before:
                        encoder_output = self.after_norm(encoder_output)
                    intermediate_outputs.append(encoder_output)
```

File: espnet/nets/pytorch_backend/conformer/encoder.py [hunk 2]
```diff
@@ -236,10 +245,25 @@ def forward(self, xs, masks):
...
+
+            if self.intermediate_layers is not None and layer_idx + 1 in self.intermediate_layers:
+                # intermediate branches also require normalization.
+                encoder_output = xs
+                if isinstance(encoder_output, tuple):
+                    encoder_output = encoder_output[0]
+                    if self.normalize_before:
+                        encoder_output = self.after_norm(encoder_output)
+                intermediate_outputs.append(encoder_output)
+
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/conformer/encoder.py:260]:
```suggestion
        if self.intermediate_layers is None:
            xs, masks = self.encoders(xs, masks)
        else:
            intermediate_outputs = []
            for layer_idx, encoder_layer in enumerate(self.encoders):
                xs, masks = encoder_layer(xs, masks)
        
                if layer_idx + 1 in self.intermediate_layers:
                    # intermediate branches also require normalization.
                    encoder_output = xs
                    if isinstance(encoder_output, tuple):
                        encoder_output = encoder_output[0]
                        if self.normalize_before:
                            encoder_output = self.after_norm(encoder_output)
                    intermediate_outputs.append(encoder_output)

```

File: espnet/nets/pytorch_backend/conformer/encoder_layer.py [hunk 3]
```diff
@@ -93,60 +96,69 @@ def forward(self, x_input, mask, cache=None):
...
-        if self.feed_forward_macaron is not None:
+        skip_layer = False
+        # with stochastic depth, residual connection `x + f(x)` becomes
+        # `x <- x + 1 / (1 - p) * f(x)` at training time.
+        stoch_layer_coeff = 1.0
+        if self.training and self.stochastic_depth_rate > 0:
+            skip_layer = torch.rand(1).item() < self.stochastic_depth_rate
+            stoch_layer_coeff = 1. / (1 - self.stochastic_depth_rate)
+
+        if not skip_layer:
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/conformer/encoder_layer.py:107]:
```suggestion
        if skip_layer:
            if cache is not None:
                x = torch.cat([cache, x], dim=1)
            if pos_emb is not None:
                return (x, pos_emb), mask
            return x, mask
```

Comment (hirofumi0810) [reply espnet/nets/pytorch_backend/conformer/encoder_layer.py:107]: With this change, the below codes can be kept original.

File: espnet/nets/pytorch_backend/transformer/encoder_layer.py [hunk 4]
```diff
@@ -68,34 +70,44 @@ def forward(self, x, mask, cache=None):
...
-        residual = x
-        if self.normalize_before:
-            x = self.norm2(x)
-        x = residual + self.dropout(self.feed_forward(x))
-        if not self.normalize_before:
-            x = self.norm2(x)
-
-        if cache is not None:
-            x = torch.cat([cache, x], dim=1)
+        skip_layer = False
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/transformer/encoder_layer.py:74]: Same as Conformer

File: espnet/nets/pytorch_backend/transformer/encoder_layer.py [hunk 5]
```diff
@@ -41,6 +41,7 @@ def __init__(
         dropout_rate,
         normalize_before=True,
         concat_after=False,
+        stochastic_depth_rate=0.,
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/transformer/encoder_layer.py:44]: Can you update docstrings?

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 6]
```diff
@@ -228,10 +246,14 @@ def forward(self, xs_pad, ilens, ys_pad):
             loss_ctc_data = None
         elif alpha == 1:
             self.loss = loss_ctc
+            if self.intermediate_ctc_weight > 0:
+                self.loss = (1 - self.intermediate_ctc_weight) * loss_ctc + self.intermediate_ctc_weight * loss_intermediate_ctc
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py:250]: Remove

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 7]
```diff
@@ -213,6 +223,14 @@ def forward(self, xs_pad, ilens, ys_pad):
...
                 self.ctc.softmax(hs_pad)
 
+            if self.intermediate_ctc_weight > 0 and self.intermediate_ctc_layers:
+                for hs_intermediate in hs_intermediates:
+                    # assuming hs_intermediates and hs_pad has same length / padding
+                    loss_inter = self.ctc(hs_intermediate.view(batch_size, -1, self.adim), hs_len, ys_pad)
+                    loss_intermediate_ctc += loss_inter
+
+                loss_intermediate_ctc /= len(self.intermediate_ctc_layers)
+
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py:233]:
```suggestion
            if self.intermediate_ctc_weight > 0:
                loss_ctc = (1 - self.intermediate_ctc_weight) * loss_ctc + self.intermediate_ctc_weight * loss_intermediate_ctc
```

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 8]
```diff
@@ -228,10 +246,14 @@ def forward(self, xs_pad, ilens, ys_pad):
...
         elif alpha == 1:
             self.loss = loss_ctc
+            if self.intermediate_ctc_weight > 0:
+                self.loss = (1 - self.intermediate_ctc_weight) * loss_ctc + self.intermediate_ctc_weight * loss_intermediate_ctc
             loss_att_data = None
             loss_ctc_data = float(loss_ctc)
         else:
             self.loss = alpha * loss_ctc + (1 - alpha) * loss_att
+            if self.intermediate_ctc_weight > 0:
+                self.loss = (1 - alpha - self.intermediate_ctc_weight) * loss_att + alpha * loss_ctc + self.intermediate_ctc_weight * loss_intermediate_ctc
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py:256]: Remove

**Comment by sw005320** [comment]
@jaesong, could you fix the CI error https://github.com/espnet/espnet/pull/3274/checks?check_run_id=2753649675#step:8:13 and respond to our comments/request?

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py (hunk 7 above)
Comment (jaesong) [reply espnet/nets/pytorch_backend/e2e_asr_transformer.py:233]: This would alter the perceived loss curve (in log or tensorboard), as it becomes weighted sum of two losses, not the original CTC loss. Is it ok?
Usually the original CTC loss becomes smaller with intermediate CTC, but the weighted sum of two losses would become larger than before (as intermediate loss >= CTC loss).

**Comment by jaesong** [comment]
Updates after 2a95bf2:

- Addressed changes suggested by @hirofumi0810 (except for `loss_ctc` on discussion).
- Implemented intermediate CTC for Mask-CTC. (`loss_ctc` may need changes after discussion)
- Added sample recipe for WSJ.
- Fixed black and flake8 issues from CI. I'm not able to run pytest right now though.

**Comment by sw005320** [comment]
Some CI errors still exist https://github.com/espnet/espnet/pull/3274/checks?check_run_id=3010796351#step:8:42

**Comment by sw005320** [comment]
@jaesong, could you fix the CI error?
You can run some tests locally by following https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md#4-unit-testing

**Comment by jaesong** [comment]
Fixed the remaining lint error. Seems like CI failure is not related to this pull request. Not sure about coverage (it shows issues on some files I haven't changed)

**Comment by sw005320** [comment]
Can you add a test to pass the intermediate ctc loss in the following files?
https://github.com/espnet/espnet/blob/master/test/test_e2e_asr_conformer.py
https://github.com/espnet/espnet/blob/master/test/test_e2e_asr_maskctc.py
https://github.com/espnet/espnet/blob/master/test/test_e2e_asr_transformer.py
This will improve the coverage and make your implementation stable.

After that, I'll merge this PR.

**Comment by jaesong** [comment]
All tests passed!

**Comment by sw005320** [comment]
Thanks!
This is a really good function.
We can also add this function to our main README.md

**Comment by danpovey** [comment]
Did you guys see any WER improvements from either intermediate CTC or stochastic depth?

**Comment by jaesong** [comment]
I'm the first author of intermediate CTC so I guess yes :)
In [our paper](https://arxiv.org/abs/2102.03216), we tried Transformer-CTC with various depth (from 6 layers to 48 layers). Intermediate CTC gives consistent improvement of WER. Stochastic Depth gives improvement if the layer is sufficiently deep (24 layers or more). Combination of both is better than either one.

**Comment by Cescfangs** [comment]
> I'm the first author of intermediate CTC so I guess yes :)
> In [our paper](https://arxiv.org/abs/2102.03216), we tried Transformer-CTC with various depth (from 6 layers to 48 layers). Intermediate CTC gives consistent improvement of WER. Stochastic Depth gives improvement if the layer is sufficiently deep (24 layers or more). Combination of both is better than either one.

@jaesong, have you seen improvement under AR model? In our case(large dataset), there was about 5% relative improvement under NAR model, no improvement under AR model though.

**Comment by D-X-Y** [comment]
@jaesong , thanks for your codes. Your implementation seems to be a soft-version of stochastic depth, `x <- x + 1 / (1 - p) * f(x)`. In the most vision applications, their implementation is the hard-version of stochastic depth, i.e., has the probability of`1-p` to use `x` and `p` to use `f(x)`. Have you tried the hard-version and how is the performance?
